File: leet-code/77. Combinations/Combinations77_3.java
package com.example;

import java.util.ArrayList;
import java.util.List;

public class Combinations77_3 {
    private int limit = 0;
    private int size = 0;

    public List<List<Integer>> combine(int n, int k) {
        limit = n;
        size = k;
        return getCombinations(new ArrayList<>(), new ArrayList<>(), 1);
    }

    private List<List<Integer>> getCombinations(List<List<Integer>> combinations, List<Integer> comb, int num) {
        if (comb.size() == size) {
            combinations.add(List.copyOf(comb));
            return combinations;
        }
        for (int i = num; i <= limit; i++) {
            comb.add(i);
            getCombinations(combinations, comb, i + 1);
            comb.remove(comb.size() - 1);
        }
        return combinations;
    }
}

File: leet-code/77. Combinations/Combinations77_3.kt
package com.example

class Combinations77_3 {
    fun combine(n: Int, k: Int): List<List<Int>> {
        fun recursion(combs: MutableList<List<Int>>, comb: MutableList<Int>, cur: Int): List<List<Int>> {
            if (comb.size == k) {
                combs.add(comb.toList())
                return combs
            }
            for (i in cur..n) {
                comb.add(i)
                recursion(combs, comb, i + 1)
                comb.removeLast()
            }
            return combs
        }

        return recursion(mutableListOf(), mutableListOf(), 1)
    }
}
